src/calculate_bucket.hpp

```diff
@@ -154,69 +159,29 @@ class F1Calculator {
         return std::make_pair(CalculateF(L), L);
     }
 
-    // Returns an evaluation of F1(L), and the metadata (L) that must be stored to evaluate F2,
-    // for 'number_of_evaluations' adjacent inputs.
-    inline std::vector<std::pair<Bits, Bits>> CalculateBuckets(
-        const Bits& start_L,
-        uint64_t number_of_evaluations)
+    // F1(x) values for x in range [first_x, first_x + n) are placed in res[].
+    // n must not be more than 1 << kBatchSizes.
+    void CalculateBuckets(uint64_t first_x, uint64_t n, uint64_t *res)
     {
-        uint16_t num_output_bits = k_;
-        uint16_t block_size_bits = kF1BlockSizeBits;
-
-        uint64_t two_to_the_k = (uint64_t)1 << k_;
-        if (start_L.GetValue() + number_of_evaluations > two_to_the_k) {
-            throw "Evaluation out of range";
-        }
-        // Counter for the first input
-        uint64_t counter = (start_L.GetValue() * (uint128_t)num_output_bits) / block_size_bits;
-        // Counter for the last input
-        uint64_t counter_end =
-            ((start_L.GetValue() + (uint128_t)number_of_evaluations + 1) * num_output_bits) /
-            block_size_bits;
-
-        std::vector<Bits> blocks;
-        uint64_t L = (counter * block_size_bits) / num_output_bits;
-        uint8_t ciphertext_bytes[kF1BlockSizeBits / 8];
+        uint64_t start = first_x * k_ / kF1BlockSizeBits;
+        // 'end' is one past the last keystream block number to be generated
+        // We first find the last used bit of the keystream, then convert it to
+        // block number and add one
+        uint64_t end = ((first_x + n) * k_ - 1) / kF1BlockSizeBits + 1;
```

Comment:
if the last bit is exactly divisible by `kF1BlockSizeBits`, this will be one too high, won't it?

would this be right?

```
uint64_t end = cdiv((first_x + n) * k_ - 1, kF1BlockSizeBits);
```

This would round it up to full blocks.

Reply:
actually, I think it should be:

```
uint64_t end = cdiv((first_x + n) * k_, kF1BlockSizeBits);
```

If this value ends up being too high, I suppose it's not the end of the world, since we just won't pull more than `n` bits from the buffer anyway. But the simpler expression we can use, the better I think.

Reply:
I did the math once more, and found that your second expression for `end` (without `- 1`) is mathematically equivalent to mine`. To put it another way, `cdiv(a, b)` can be defined as `(a + b - 1) / b` or `(a - 1) / b + 1`, assuming `a` is always positive.

I agree your version is a bit simpler, so I'll update the expression.

Reply:
you're right. in fact, would could defined `cdiv()` to use the `(a - 1) / b + 1` form if we believe it's better. It has less risk of overflow for instance, and maybe the arithmetic could be faster too, since it's just immediate values/constants.

Reply:
I've now pushed a new version with `cdiv` call. Regarding the implementation of `cdiv()` itself, I think the current one is better. The other version will not work with `a == 0`, and actually performs more operations than the current one when `b` is constant (3 vs 2 arithmetic ops).

The compilers also do a further optimization when `b` is a power of two, replacing division with right shift.

For these reasons I think it's preferable to keep the current version.